training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala

```diff
@@ -175,6 +176,73 @@ object FullRankLinearTrainer {
     .collect
   }
 
+  def hingeGradient(sc : SparkContext,
+                    options : FullRankLinearTrainerOptions,
+                    model : FullRankLinearModel,
+                    pointwise : RDD[Example]) : Array[((String, String), GradientContainer)] = {
+    val modelBC = sc.broadcast(model)
+
+    pointwise
+      .mapPartitions(partition => {
+        val model = modelBC.value
+        val labelToIdx = model.getLabelToIndex()
+        val dim = model.getLabelDictionary.size()
+        val gradient = scala.collection.mutable.HashMap[(String, String), GradientContainer]()
+        val weightVector = model.getWeightVector()
+        val rnd = new Random()
+
+        partition.foreach(examples => {
+          val flatFeatures = Util.flattenFeature(examples.example.get(0))
+          val labels = flatFeatures.get(options.rankKey)
+          if (labels != null && labels.size() > 0) {
+            val posLabels = labels.toArray
+            // Pick a random positive label
+            val posLabelRnd = rnd.nextInt(posLabels.size)
+            val (posLabel, posMargin) = posLabels(posLabelRnd)
+            val posIdx = labelToIdx.get(posLabel)
+            // Pick a random other label
+            var negIdx = rnd.nextInt(dim)
+            while (negIdx == posIdx) {
+              negIdx = rnd.nextInt(dim)
+            }
+            val negLabel = model.getLabelDictionary.get(negIdx).label
+            val negMargin : Double = if (labels.containsKey(negLabel)) labels.get(negLabel) else 0.0
```

Comment:
I am very confused here. I thought for one example, labels would contain positive labels for this example